examples/hamiltonian_representation.py

```diff
@@ -0,0 +1,135 @@
+import math
+from typing import Dict, Tuple
+
+from sympy.logic.boolalg import And, Not, Or, Xor
+from sympy.core.symbol import Symbol
+
+import cirq
+
+# References:
+# [1] On the representation of Boolean and real functions as Hamiltonians for quantum computing
+#     by Stuart Hadfield
+# [2] https://www.youtube.com/watch?v=AOKM9BkweVU is a useful intro
+# [3] https://github.com/rsln-s/IEEE_QW_2020/blob/master/Slides.pdf
+
+
+class HamiltonianList:
+    """A container class of Boolean function as equation (2) or [1]"""
+
+    def __init__(self, hamiltonians: Dict[Tuple[int, ...], float]):
+        # The representation is Tuple[int, ...] to weights. The tuple contains the integers of
+        # where Z_i is present. For example, Z_0.Z_3 would be (0, 3), and I is the empty tuple.
+        self.hamiltonians = {h: w for h, w in hamiltonians.items() if math.fabs(w) > 1e-12}
+
+    def __str__(self):
+        # For run-to-run identicalness, we sort the keys lexicographically.
+        return "; ".join(
+            f"{self.hamiltonians[h]:.2f}.{'.'.join('Z_%d' % d for d in h) if h else 'I'}"
+            for h in sorted(self.hamiltonians)
+        )
+
+    def __add__(self, other):
+        return self._signed_add(other, 1.0)
+
+    def __sub__(self, other):
+        return self._signed_add(other, -1.0)
+
+    def _signed_add(self, other, sign: float):
+        hamiltonians = self.hamiltonians.copy()
+        for h, w in other.hamiltonians.items():
+            if h not in hamiltonians:
+                hamiltonians[h] = 0
+            hamiltonians[h] += sign * w
+        return HamiltonianList(hamiltonians)
+
+    def __rmul__(self, other: float):
+        return HamiltonianList({k: other * w for k, w in self.hamiltonians.items()})
+
+    def __mul__(self, other):
+        hamiltonians = {}
+        for h1, w1 in self.hamiltonians.items():
+            for h2, w2 in other.hamiltonians.items():
+                h = tuple(sorted(set(h1).symmetric_difference(h2)))
+                w = w1 * w2
+                if h not in hamiltonians:
+                    hamiltonians[h] = 0
+                hamiltonians[h] += w
+        return HamiltonianList(hamiltonians)
+
+    @staticmethod
+    def O():
+        return HamiltonianList({})
+
+    @staticmethod
+    def I():
+        return HamiltonianList({(): 1.0})
+
+    @staticmethod
+    def Z(i: int):
+        return HamiltonianList({(i,): 1.0})
+
+
+def build_hamiltonian_from_boolean(boolean_expr, name_to_id) -> HamiltonianList:
+    """Builds the Hamiltonian representation of Boolean expression as per [1]:
+
+    Args:
+        boolean_expr: A Sympy expression containing symbols and Boolean operations
+        name_to_id: A dictionary from symbol name to an integer, typically built by calling
+            get_name_to_id().
+
+    Return:
+        The HamiltonianList that represents the Boolean expression.
+    """
+
+    if isinstance(boolean_expr, (And, Not, Or, Xor)):
+        sub_hamiltonians = [
```

Comment:
Is it true that `sub_hamiltonians` always has a length 1 or 2? If yes, I think the later construction can be simplified

Reply:
Not with my current version of Sympy. For `test_build_hamiltonian_from_boolean` with input `x0 ^ x1 ^ x2` the XOR has three operands. I think we could try to do things recursively, of course, but I am not sure it would make the code cleaner.

Reply:
Aha, you are right. I am not familiar with sympy_parser. Then, current style is good.